From f2f5c491b55d63bfe9ee9c34329aeaefd9d65a30 Mon Sep 17 00:00:00 2001
From: Arzhan Kinzhalin <arzhan.i.kinzhalin@intel.com>
Date: Tue, 19 Feb 2019 09:38:48 -0800
Subject: [PATCH 1/2] Fix size check.

---
 src/bootman/kernel.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bootman/kernel.c b/src/bootman/kernel.c
index 95488f2..63fb2b1 100644
--- a/src/bootman/kernel.c
+++ b/src/bootman/kernel.c
@@ -449,7 +449,7 @@ bool cbm_parse_system_kernel(const char *inp, SystemKernel *kernel)
                 ++len;
         }
 
-        if (len < 1 || len >= CBM_KELEM_LEN) {
+        if (len < 1 || len > CBM_KELEM_LEN) {
                 return false;
         }
 
-- 
2.20.1

